Source/WebCore/html/HTMLFormControlElement.cpp

```diff
@@ -321,6 +322,80 @@ String HTMLFormControlElement::resultForDialogSubmit() const
     return attributeWithoutSynchronization(HTMLNames::valueAttr);
 }
 
+static const AtomString& toggleAtom()
+{
+    static MainThreadNeverDestroyed<const AtomString> identifier("toggle"_s);
+    return identifier;
+}
+
+static const AtomString& showAtom()
+{
+    static MainThreadNeverDestroyed<const AtomString> identifier("show"_s);
+    return identifier;
+}
+
+static const AtomString& hideAtom()
+{
+    static MainThreadNeverDestroyed<const AtomString> identifier("hide"_s);
+    return identifier;
+}
+
+// https://html.spec.whatwg.org/#popover-target-element
+HTMLElement* HTMLFormControlElement::popoverTargetElement() const
+{
+    auto canInvokePopovers = [](const HTMLFormControlElement& element) -> bool {
+        if (auto* inputElement = dynamicDowncast<HTMLInputElement>(element))
+            return inputElement->isTextButton() || inputElement->isImageButton();
+        return is<HTMLButtonElement>(element);
+    };
+
+    if (!canInvokePopovers(*this))
+        return nullptr;
+
+    if (isDisabledFormControl())
+        return nullptr;
+
+    if (form() && isSubmitButton())
+        return nullptr;
+
+    auto* element = dynamicDowncast<HTMLElement>(getElementAttribute(popovertargetAttr));
+    if (element && element->popoverState() != PopoverState::None)
+        return element;
+    return nullptr;
+}
+
+const AtomString& HTMLFormControlElement::popoverTargetAction() const
+{
+    auto value = attributeWithoutSynchronization(HTMLNames::popovertargetactionAttr);
+
+    if (equalIgnoringASCIICase(value, showAtom()))
```

Comment:
Does this really need to be case insensitive?

Reply:
It is case insensitive in Chrome, popovertargetaction="SHow" seems to work there.

Reply:
Enumerated attribute values are always case-insensitive.